src/test/java/pages/BasePage.java

```diff
@@ -0,0 +1,39 @@
+package pages;
+
+import io.github.bonigarcia.wdm.WebDriverManager;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Parameters;
+import java.time.Duration;
+
+public class BasePage {
+    protected static WebDriver driver;
+    protected static WebDriverWait wait;
+    protected static Actions actions;
+    @BeforeSuite
+    static void setupClass() {
+        WebDriverManager.chromedriver().setup();
+    }
+
+    @BeforeMethod
+    @Parameters({"BaseURL"})
+    public void setUpBrowser(String BaseURL)  {
+        driver = new ChromeDriver();
+        wait = new WebDriverWait(driver,Duration.ofSeconds(10));
+        actions = new Actions(driver);
+        String url = BaseURL;
+        driver.get(url);
+    }
+    @AfterMethod
+    public void tearDown() {
+        driver.quit();
+    }
```

Comment:
Our **BasePage** class will have a constructor which takes a WebDriver object to initialize a WebDriverWait object.

For example 

```
public class BasePage {
    protected static WebDriver driver;
    protected static WebDriverWait wait;
    protected static Actions actions;

    public BasePage(WebDriver givenDriver) {
        driver = givenDriver;
        wait = new WebDriverWait(driver, Duration.ofSeconds(5));
    }

    public WebElement findElement(By locator) {
        return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

    public void doubleClick(By locator) {
        actions.doubleClick(findElement(locator)).perform();
    }

    public void click(By locator) {
        wait.until(ExpectedConditions.elementToBeClickable(locator));
    }

    public void contextClick(By locator) {
        actions.contextClick(findElement(locator)).perform();
    }
}
```


**BaseTest** class holds the methods to initialize and terminate the WebDriver object.

**Please, remove code redundancies**

Don't repeat yourself! ( In Java, it means don't write the same code repeatedly. )
 

![image](https://user-images.githubusercontent.com/104761482/218390583-f70aa117-4bb1-4b08-8b01-5dd51b18bffa.png)

Reply:
Ok but how to start Chrome Browser if we don't have 
 
@BeforeSuite
    static void setupClass() {
        WebDriverManager.chromedriver().setup();
    }

    @BeforeMethod
    public void setUpBrowser() {
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--disable-notifications");
        driver = new ChromeDriver(options);
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
        wait = new WebDriverWait(driver, Duration.ofSeconds(4));
       // driver.manage().window().maximize();
        driver.get(url);
    }

    @AfterMethod
    public void tearDown() {
        driver.quit();
    }